homeassistant/components/tuya/__init__.py

```diff
@@ -260,18 +163,71 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
     return unload
 
 
-async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Async setup hass config entry."""
     _LOGGER.debug("tuya.__init__.async_setup_entry-->%s", entry.data)
 
-    hass.data[DOMAIN] = {
-        TUYA_HA_TUYA_MAP: {},
-        TUYA_HA_DEVICES: [],
-        TUYA_SETUP_PLATFORM: set(),
-    }
+    hass.data[DOMAIN] = {TUYA_HA_TUYA_MAP: {}, TUYA_HA_DEVICES: {}}
 
     success = await _init_tuya_sdk(hass, entry)
     if not success:
         return False
 
     return True
+
+
+class DeviceListener(TuyaDeviceListener):
+    """Device Update Listener."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Init DeviceListener."""
+
+        self.hass = hass
+
+    def update_device(self, device: TuyaDevice) -> None:
+        """Update device status."""
+        if device.id in self.hass.data[DOMAIN][TUYA_HA_DEVICES]:
+            _LOGGER.debug(
+                "_update-->%s;->>%s",
+                self,
+                self.hass.data[DOMAIN][TUYA_HA_DEVICES][device.id].tuya_device.status,
+            )
+            async_dispatcher_send(self.hass, TUYA_HA_SIGNAL_UPDATE_ENTITY)
```

Comment:
Include the device id in this signal to let the dispatch helper look up the correct signal handler fast.
```suggestion
            async_dispatcher_send(self.hass, f"{TUYA_HA_SIGNAL_UPDATE_ENTITY}_{device.id}")
```